### PR TITLE
use github action to deploy to test.pypi.org and pypi.org

### LINKED ISSUE #929: BLD: Use Azure Pipelines for deploying new tagged releases
**Is your feature request related to a problem? Please describe.**
It would be ideal for Azure Pipelines (AP) to execute the upload to PyPI on new tag versions for Continuous Delivery (CD). This will allow us to eventually drop the use of Travis CI so we have one less config file/environment to manage.

**Describe the solution you'd like**
Implement AP's functionality with the Python package `twine` to execute this on a new tag. 

**Describe alternatives you've considered**
It could be possible to create a GitHub Action to do this however this would add another platform to our CI while trying to remove Travis.

**Additional context**
Travis additionally manages:
* [ ] upload of coverage data to coveralls
* [ ] testing of Python3.5-min

These will also need to be transferred to AP before we can drop Travis.

### PR DESCRIPTION
**Pull request by wholmgren**
- [x] Closes #929 
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

I started by trying to add to the Azure CI so that we didn't have to worry about yet another CI service but the Azure documentation on this is just terrible. This workflow is largely modeled on

- [PyPA example](https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/)
- [pypi-publish action documentation](https://github.com/marketplace/actions/pypi-publish)
- [metpy publish.yml](https://github.com/Unidata/MetPy/blob/master/.github/workflows/release.yml)

A nice change here is that all commits will be pushed to test.pypi.org so we can ensure that the distribution is always built as expected. This also gives users another way to get prerelease versions of pvlib at https://test.pypi.org/project/pvlib/

By default, github would also run this action on pushes to forks. None of the attempts to post would be successful because they don't have the secret I added to https://github.com/pvlib/pvlib-python/settings/secrets. But I don't want to unnecessarily clutter users notifications with github action failures, so I added a check `if github.repository == 'pvlib/pvlib-python'`.

See [this build](https://github.com/wholmgren/pvlib-python/runs/1053509563?check_suite_focus=true) to confirm that the action builds a wheel and sdist with the right version.

See [this build](https://github.com/wholmgren/pvlib-python/runs/1053530435?check_suite_focus=true) to confirm that the `if github.repository` condition works (or at least prevents running on my fork).

I removed the ability for the Travis configuration to deploy the releases. I think this makes the Travis configuration obsolete, but I want to wait to delete it until we know this new workflow works.

### CODE REVIEW

**Comment by wholmgren** [comment]
@mikofski @kanderso-nrel @cwhanse can you review this if you have experience with these tools?

File: .github/workflows/publish-to-pypi-testpypi.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+jobs:
+  build-n-publish:
+    name: Build and publish distributions to PyPI (releases only) and TestPyPI (all commits)
+    if: github.repository == 'pvlib/pvlib-python'
+    runs-on: ubuntu-latest
+    steps:
+    # fetch all commits and tags so versioneer works
+    - uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
```
Comment (kanderso-nrel) [inline .github/workflows/publish-to-pypi-testpypi.yml:14]: Fetching all of history seems a little overkill -- is it worth it to copy [metpy's approach](https://github.com/Unidata/MetPy/blob/master/.github/workflows/release.yml#L17)?  Or maybe fetching is cheap so let's not worry about it.

File: .github/workflows/publish-to-pypi-testpypi.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    - name: Build packages
+      run: python setup.py sdist bdist_wheel
+
+    - name: Publish distribution to Test PyPI
+      uses: pypa/gh-action-pypi-publish@master
+      with:
+        user: __token__
+        password: ${{ secrets.test_pypi_password }}
+        repository_url: https://test.pypi.org/legacy/
```
[the remark this answers is not in the document]
Comment (wholmgren) [reply .github/workflows/publish-to-pypi-testpypi.yml:34]: I think the best reason to use Test PyPI is to ensure that your distribution workflow (manual or automated) works as you expect it to.

It's hard for me to imagine a situation in which we'd run into the race condition and not want to be alerted to a problem for manual fixes. The action only works on pushes to `pvlib/pvlib-python` - it doesn't run on PRs.

File: .github/workflows/publish-to-pypi-testpypi.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    - name: Install build tools
+      run: |
+        python -m pip install --upgrade pip
+        python -m pip install --upgrade setuptools wheel
+
+    - name: Build packages
+      run: python setup.py sdist bdist_wheel
+
+    - name: Publish distribution to Test PyPI
+      uses: pypa/gh-action-pypi-publish@master
```
[the remark this answers is not in the document]
Comment (wholmgren) [reply .github/workflows/publish-to-pypi-testpypi.yml:30]: I was wondering the same thing when I copied it from the PyPA guide. That guide also used master for the other actions that we're now pinning (in part copying metpy and other examples). Either way is fine with me.

File: .github/workflows/publish-to-pypi-testpypi.yml (hunk 1 above)
Comment (wholmgren) [reply .github/workflows/publish-to-pypi-testpypi.yml:14]: It doesn't take more than a few seconds for this to run so I think we should keep it simple for now.

File: .github/workflows/publish-to-pypi-testpypi.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (kanderso-nrel) [reply .github/workflows/publish-to-pypi-testpypi.yml:34]: Oh, I mistakenly thought that `push` included pushes to PRs.  I'm not 100% satisfied with the documentation on this, but it seems that you are correct that it has to be a push to `pvlib/pvlib-python` specifically: https://developer.github.com/v3/activity/event_types/#pushevent

[the remark this answers is not in the document]
Comment (wholmgren) [reply .github/workflows/publish-to-pypi-testpypi.yml:34]: To be clear, the action does run on pushes to forks, but it stops before doing anything meaningful. See screen shot below for a recent history of the actions on my fork.

Screen Shot 2020-08-31 at 7 47 03 PM

Unfortunately it does not seem that there is a way for us to disable actions on forks, but forks may choose to disable actions in https://github.com/{username}/pvlib-python/settings/actions

**Comment by wholmgren** [comment]
Thanks for the feedback @kanderso-nrel. Let's see how this works...

**Comment by wholmgren** [comment]
Fails because pypi won't accept PEP 440 local versions. https://github.com/pvlib/pvlib-python/runs/1058055206?check_suite_focus=true

Follow up PR to restrict to tags incoming.

**Pull request merged**